### PR TITLE
[TOD] Core conversation structure, serialization, const tokens

### PR DESCRIPTION
**Pull request by moyapchen**
Core components. How they are used + testing gets done later in the stack.

### CODE REVIEW

File: parlai/core/tod/tod_core.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+
+@dataclass
+class TodStructuredEpisode:
+    """
+    Dataclass for episode-level data.
+    """
+
+    # Variables set by those using this class
+    delex: bool = False  # Set to true and this class will handle delexicalizing call + response utterances based on API calls and responses exposed to this class.
```
Comment (moyapchen) [inline parlai/core/tod/tod_core.py:95]: Oof, need to find a linter that will wrap comments hmm...

Docformatter also fails to format the other comments inside of """'s. Playing around with it, not quite sure why (CC: @stephenroller )

**Review by EricMichaelSmith — COMMENTED** [review]
Seems generally reasonable to me with my limited knowledge of the context - a few minor comments. Will defer to others for approval

File: parlai/core/tod/tod_core.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Task Oriented Dialogue (TOD) enums and base classes.
+
+This file defines standard tokens, classes for round and conversation structure, and a serialization class to aid in converting between these.
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/tod_core.py:10]: Good callout. I should probably define that before using it in this class.

File: parlai/core/tod/tod_core.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+class TodAgentType(str, Enum):
+    USER_UTT_AGENT = "user_utt_model"
+    API_CALL_AGENT = "api_call_model"
+    API_RESP_AGENT = "api_resp_model"
+    SYSTEM_UTT_AGENT = "system_utt_model"
+    API_SCHEMA_GROUNDING_AGENT = "api_schema_grounding_model"
+    GOAL_GROUNDING_AGENT = "goal_grounding_model"
+
+
+TOD_AGENT_TYPE_TO_PREFIX = {
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/tod_core.py:45]: Yeah, I end up using it in files external to this one later in the stack. (In hindsight, they're all one-to-one mappings so maybe I didn't exactly need this extra level of indirection, but I do think separating out "Agent Type" from the "Token that Agent Type uses" makes sense.)

File: parlai/core/tod/tod_core.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    TodAgentType.SYSTEM_UTT_AGENT: STANDARD_SYSTEM_UTTERANCE,
+    TodAgentType.API_SCHEMA_GROUNDING_AGENT: STANDARD_API_SCHEMAS,
+    TodAgentType.GOAL_GROUNDING_AGENT: STANDARD_GOAL,
+}
+
+
+@dataclass
+class TodStructuredRound:
+    """
+    Dataclass for rounds.
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/tod_core.py:58]: Will update diff with comment here. :)

**Review by stephenroller — APPROVED** [review]
accepting eric's approval as final approval

**Review by EricMichaelSmith — APPROVED** [review]
Approving, thanks for the responses!

**Comment by moyapchen** [comment]
Thanks for the reviews!

**Pull request merged**